ADR/0027-availability-probe-framework.md

```diff
@@ -0,0 +1,125 @@
+API server. It generates
+[Jobs](https://github.com/kubernetes/kube-state-metrics/blob/main/docs/job-metrics.md) and
+[CronJob](https://github.com/kubernetes/kube-state-metrics/blob/main/docs/cronjob-metrics.md)
+metrics that will be processed using Prometheus recording rules in order to
+generate AppStudio's availability Prometheus metric.
+
+A single set of rules will be defined globally which will apply for all CronJobs.
+
+The resulting Prometheus metric will identify each probe based on its CronJob name and the namespace in which the CronJob was defined:
+
+* The name of the CronJob will be translated to a Prometheus metric label denoting the
+  name of the probe.
+* The namespace in which the CronJob is defined will be translated to a Prometheus
+  metric label denoting the probe's namespace.
+
+#### Probes Requirements and Design Considerations
+
+A Job running for a probe is required to:
+
+* Evaluate the availability of the component being probed.
+* Exit with status `Failed` if the component was evaluated to be unavailable or with
+  status `Complete` if it was evaluated to be available (referring to the status field
+  of the Kubernetes Job resource).
+* Be able to exhaust its `backoffLimit` by the time it's due to run again (e.g. if the
+  cronjob is defined to run every 10 minutes and can take up to 2 minutes to execute
+  it cannot have a `backoffLimit` larger than 4).
+* Clean up all resources generated during its run.
+
+Considerations for defining probes' CronJobs:
+
+* Where should it run in order to provide reliable evaluation?
+    * Which namespace?
+    * Which clusters?
+* What sort of permissions does it require to have?
+    * Would that be a good-enough representation of what it aims to evaluate?
+* What sort of APIs does it need to access?
+* How often should it run?
+* Does it affect performance in a reasonable manner?
+* How resources are to be cleaned up?
+    * Upon startup? completion? failure?
+    * Using another CronJob? Finalizers?
+
+#### Naming Convention
+
+To allow generating the Prometheus metric only for relevant CronJobs, probe CronJob
+names should have a standardized format:
+
+* `appstudio-probe-<probe_name>`
+
+`appstudio-probe-` being a literal to be used in order to capture only the relevant
+CronJobs, and `<probe_name>` is a variable to be translated to a label in the resulting
+Prometheus metric that will correlate the value to the individual probe or check.
+
+To allow aggregating the Prometheus availability metric per component, the namespaces
+in which the CronJobs will be created should have a standardized format:
+
+* `appstudio-probe-<component_name>`
+
+`<component_name>` is the name of the component under which probes will be aggregated.
+
+The nature and size of each service and component will dictate the number of probes it
+should have and how they should be aggregated. E.g. some service may have multiple
```

Comment:
```suggestion
should have. E.g. some service may have multiple
```
It shouldn't change _how_ they get aggregated. The planned `PrometheusRules` should be capable of aggregation for a component that uses 100 probes in the same way as if there is just 1 probe.

Reply:
done